clang/lib/AST/PreorderAST.cpp

```diff
@@ -188,53 +362,94 @@ bool PreorderAST::CompareNodes(const Node *N1, const Node *N2) {
       return Lex.CompareExpr(L1->E, L2->E) == Result::LessThan;
     }
-
-    // N2:OperatorNodeExpr < N1:LeafExprNode.
-    return false;
   }
 
-  // N1:OperatorNodeExpr < N2:LeafExprNode.
-  if (isa<LeafExprNode>(N2))
-    return true;
-
-  // Compare N1:OperatorNode and N2:OperatorNode.
-  const auto *O1 = dyn_cast<OperatorNode>(N1);
-  const auto *O2 = dyn_cast<OperatorNode>(N2);
-
-  if (O1->Opc != O2->Opc)
-    return O1->Opc < O2->Opc;
-  return O1->Children.size() < O2->Children.size();
+  return true;
 }
 
 void PreorderAST::Sort(Node *N) {
-  auto *O = dyn_cast_or_null<OperatorNode>(N);
-  if (!O)
+  if (!N)
     return;
 
-  // Sort the children first.
-  for (auto *Child : O->Children)
-    if (isa<OperatorNode>(Child))
-      Sort(Child);
+  switch (N->Kind) {
+    case Node::NodeKind::OperatorNode: {
+      auto *O = dyn_cast<OperatorNode>(N);
 
-  // We can only sort if the operator is commutative and associative.
-  if (!O->IsOpCommutativeAndAssociative())
-    return;
+      // Sort the children first.
+      for (auto *Child : O->Children)
+        Sort(Child);
 
-  // Sort the children.
-  llvm::sort(O->Children.begin(), O->Children.end(),
-             [&](const Node *N1, const Node *N2) {
-               return CompareNodes(N1, N2);
-            });
+      // We can only sort if the operator is commutative and associative.
+      if (!O->IsOpCommutativeAndAssociative())
+        return;
+
+      // Sort the children.
+      llvm::sort(O->Children.begin(), O->Children.end(),
+                [&](const Node *N1, const Node *N2) {
+                  return CompareNodes(N1, N2);
+                });
+      break;
+    }
+    case Node::NodeKind::UnaryOperatorNode: {
+      auto *U = dyn_cast<UnaryOperatorNode>(N);
+      Sort(U->Child);
+      break;
+    }
+    case Node::NodeKind::MemberNode: {
+      auto *M = dyn_cast<MemberNode>(N);
+      Sort(M->Base);
+      break;
+    }
+    case Node::NodeKind::ImplicitCastNode: {
+      auto *I = dyn_cast<ImplicitCastNode>(N);
+      Sort(I->Child);
+      break;
+    }
+    default:
+      break;
+  }
 }
 
 void PreorderAST::ConstantFold(Node *N, bool &Changed) {
+  if (Error)
+    return;
+
+  if (!N)
+    return;
+
+  switch (N->Kind) {
+    case Node::NodeKind::OperatorNode: {
+      auto *O = dyn_cast<OperatorNode>(N);
+      ConstantFoldOperator(O, Changed);
+      break;
+    }
+    case Node::NodeKind::UnaryOperatorNode: {
+      auto *U = dyn_cast<UnaryOperatorNode>(N);
+      ConstantFold(U->Child, Changed);
+      break;
+    }
+    case Node::NodeKind::MemberNode: {
+      auto *M = dyn_cast<MemberNode>(N);
+      ConstantFold(M->Base, Changed);
+      break;
+    }
+    case Node::NodeKind::ImplicitCastNode: {
+      auto *I = dyn_cast<ImplicitCastNode>(N);
+      ConstantFold(I->Child, Changed);
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+void PreorderAST::ConstantFoldOperator(OperatorNode *O, bool &Changed) {
```

Comment:
We do not need to change the name of this function. It can be called as `ConstantFold` and C++ function overloading would figure out that this function needs to be called based on the type of the first argument (which would be `Node *` in one function and `OperatorNode *` in the other)

Reply:
I tried renaming `ConstantFoldOperator` to `ConstantFold` and unfortunately constant folding no longer worked. It seems that `ConstantFold(BinaryOperatorNode *B)` will only be called on a `Node * N` if `N` is declared as a `BinaryOperatorNode *`. For example, I tried changing the type of `PreorderAST.Root` from `Node *` to `BinaryOperatorNode *` and then `ConstantFold(Root)` called the correct function; however, this does not work for any descendants of a `BinaryOperatorNode`. For example, the expression `*(*(p + 1 + 2) + 3 + 4)` should be constant folded to `*(*(p + 3) + 7)`, but it is not.

Reply:
Discussed offline - the need to have two different `ConstantFold` and `ConstantFoldOperator` methods should no longer be a consideration after the follow-up work in issue #1032 to move methods such as `ConstantFold` to their respective `Node` classes. For now, the methods `ConstantFold` and `ConstantFoldOperator` can keep their respective names.